homeassistant/components/alphaess/__init__.py

```diff
@@ -0,0 +1,84 @@
+"""The AlphaEss integration."""
+from __future__ import annotations
+
+# import asyncio
+from datetime import timedelta
+import logging
+
+from alphaess import alphaess
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+
+# from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN
+
+# TODO List the platforms that you want to support.
+# For your initial PR, limit it to 1 platform.
+PLATFORMS = ["sensor"]
+
+SCAN_INTERVAL = timedelta(minutes=5)
+
+_LOGGER: logging.Logger = logging.getLogger(__package__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Alpha ESS from a config entry."""
+
+    if hass.data.get(DOMAIN) is None:
+        hass.data.setdefault(DOMAIN, {})
+
+    client = alphaess.alphaess()
+    client.username = entry.data.get("username")
+    client.password = entry.data.get("password")
+
+    coordinator = AlphaESSDataUpdateCoordinator(hass, client=client)
+    await coordinator.async_config_entry_first_refresh()
+
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+
+    entry.async_on_unload(entry.add_update_listener(async_update_options))
+
+    return True
+
+
+class AlphaESSDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching data from the API."""
+
+    def __init__(self, hass: HomeAssistant, client: alphaess.alphaess) -> None:
+        """Initialize."""
+        self.api = client
+        self.platforms = ["sensor"]
+        self.update_method = self._async_update_data
+
+        super().__init__(hass, _LOGGER, name=DOMAIN, update_interval=SCAN_INTERVAL)
+
+    async def _async_update_data(self):
+        """Update data via library."""
+        try:
+            _LOGGER.info("Trying to query AlphaESS api data")
+            return await self.api.getdata()
+        except Exception as exception:
```

Comment:
Do you have specific exceptions in your client class? Or does it always throw a generic exception, where it can recover from if you try again?

Reply:
The client class is not yet very sophisticated.  I know it needs more work to get to a state where the home assistant integration can move up the quality scale.

Reply:
Those broad exceptions aren't allowed most of the time, so I would advise to see if you can catch errors like TimeOuts or wrong credentials asap.

Reply:
Taken me too long to come back to this but I have updated the client class to identify connection errors and authentication problems separately.